src/main/java/org/javarosa/xpath/expr/ExpressionCacheKey.java

```diff
@@ -10,22 +16,64 @@
 
     private InFormCacheableExpr expr;
 
-    public ExpressionCacheKey(InFormCacheableExpr expr) {
+    // These are each only relevant to caching in some cases, and will be null in a cache key for
+    // which they are not relevant
+    @Nullable
+    private TreeReference contextRef;
+    @Nullable
+    private TreeReference originalContextRef;
+
+    public ExpressionCacheKey(InFormCacheableExpr expr, EvaluationContext ec) {
         this.expr = expr;
+        for (int contextType : getRelevantContextTypes(expr)) {
```

Comment:
It seems a bit tenuous to me that there's no way to make sure that this evaluation (getRelevantContextTypes) is being cached (along the lifecycle of the expr object rather than the lifecycle of this key object) . Is it just too fast to matter?

Reply:
The final version of this has changed, and it is cached there.